src/machines/systemIO/systemIOMachine.ts

```diff
@@ -0,0 +1,448 @@
+        }
+      }): Promise<{ value: boolean; error: unknown }> => {
+        return { value: true, error: undefined }
+      }
+    ),
+  },
+}).createMachine({
+  initial: SystemIOMachineStates.idle,
+  // Remember, this machine and change its projectDirectory at any point
+  // '' will be no project directory, aka clear this machine out!
+  // To be the absolute root of someones computer we should take the string of path.resolve() in node.js which is different for each OS
+  context: () => ({
+    folders: [],
+    defaultProjectFolderName: DEFAULT_PROJECT_NAME,
+    projectDirectoryPath: NO_PROJECT_DIRECTORY,
+    hasListedProjects: false,
+    requestedProjectName: { name: NO_PROJECT_DIRECTORY },
+    requestedFileName: {
+      project: NO_PROJECT_DIRECTORY,
+      file: NO_PROJECT_DIRECTORY,
+    },
+    canReadWriteProjectDirectory: { value: true, error: undefined },
+    clearURLParams: { value: false },
+  }),
+  states: {
+    [SystemIOMachineStates.idle]: {
+      on: {
+        // on can be an action
+        [SystemIOMachineEvents.readFoldersFromProjectDirectory]: {
+          target: SystemIOMachineStates.readingFolders,
+        },
+        [SystemIOMachineEvents.setProjectDirectoryPath]: {
+          target: SystemIOMachineStates.checkingReadWrite,
+          actions: [SystemIOMachineActions.setProjectDirectoryPath],
+        },
+        [SystemIOMachineEvents.navigateToProject]: {
+          actions: [SystemIOMachineActions.setRequestedProjectName],
+        },
+        [SystemIOMachineEvents.navigateToFile]: {
+          actions: [SystemIOMachineActions.setRequestedFileName],
+        },
+        [SystemIOMachineEvents.createProject]: {
+          target: SystemIOMachineStates.creatingProject,
+        },
+        [SystemIOMachineEvents.renameProject]: {
+          target: SystemIOMachineStates.renamingProject,
+        },
+        [SystemIOMachineEvents.deleteProject]: {
+          target: SystemIOMachineStates.deletingProject,
+        },
+        [SystemIOMachineEvents.createKCLFile]: {
+          target: SystemIOMachineStates.creatingKCLFile,
+        },
+        [SystemIOMachineEvents.setDefaultProjectFolderName]: {
+          actions: [SystemIOMachineActions.setDefaultProjectFolderName],
+        },
+        [SystemIOMachineEvents.importFileFromURL]: {
+          target: SystemIOMachineStates.importFileFromURL,
+        },
+      },
+    },
+    [SystemIOMachineStates.readingFolders]: {
+      invoke: {
+        id: SystemIOMachineActors.readFoldersFromProjectDirectory,
+        src: SystemIOMachineActors.readFoldersFromProjectDirectory,
+        input: ({ context }) => {
+          return context
+        },
+        onDone: {
+          target: SystemIOMachineStates.idle,
+          actions: [
+            SystemIOMachineActions.setFolders,
+            assign({ hasListedProjects: true }),
+          ],
+        },
+        onError: {
+          target: SystemIOMachineStates.idle,
```

Comment:
Failing to read the folders doesn't toast any sort of error to the user?

Reply:
As of right now, no. I don't exactly know what the best way to handle this is from the UX experience.

Reply:
Alright just something we should keep an eye on. Maybe a toast or a error log could help diagnose if this error ever comes up for a user.